infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesClientFactory.java

```diff
@@ -63,7 +80,78 @@ public KubernetesClientFactory(
     if (doTrustCerts != null) {
       configBuilder.withTrustCerts(doTrustCerts);
     }
-    this.client = new UnclosableKubernetesClient(configBuilder.build());
+
+    Config config = configBuilder.build();
+    return config;
+  }
+
+  protected OkHttpClient getHttpClient() {
+    return httpClient;
+  }
+
+  protected Config getDefaultConfig() {
+    return defaultConfig;
+  }
+
+  public Config buildConfig(Config defaultConfig, @Nullable String workspaceId)
+      throws InfrastructureException {
+    return defaultConfig;
+  }
+
+  /**
+   * Creates instance of {@link KubernetesClient}.
+   *
+   * @throws InfrastructureException if any error occurs on client instance creation.
+   */
+  private KubernetesClient create(Config config) throws InfrastructureException {
+    OkHttpClient clientHttpClient =
+        httpClient.newBuilder().authenticator(Authenticator.NONE).build();
+    OkHttpClient.Builder builder = clientHttpClient.newBuilder();
+    builder.interceptors().clear();
+    clientHttpClient =
+        builder
+            .addInterceptor(
+                new Interceptor() {
+                  @Override
+                  public Response intercept(Chain chain) throws IOException {
+                    Request request = chain.request();
+                    if (isNotNullOrEmpty(config.getUsername())
+                        && isNotNullOrEmpty(config.getPassword())) {
+                      Request authReq =
+                          chain
+                              .request()
+                              .newBuilder()
+                              .addHeader(
+                                  "Authorization",
+                                  Credentials.basic(config.getUsername(), config.getPassword()))
+                              .build();
+                      return chain.proceed(authReq);
+                    } else if (isNotNullOrEmpty(config.getOauthToken())) {
+                      Request authReq =
+                          chain
+                              .request()
+                              .newBuilder()
+                              .addHeader("Authorization", "Bearer " + config.getOauthToken())
+                              .build();
+                      return chain.proceed(authReq);
+                    }
+                    return chain.proceed(request);
+                  }
+                })
+            .build();
+
+    return new UnclosableKubernetesClient(clientHttpClient, config);
+  }
+
+  /**
+   * Creates instance of {@link KubernetesClient}.
+   *
+   * @throws InfrastructureException if any error occurs on client instance creation.
+   */
+  public KubernetesClient create(String workspaceId) throws InfrastructureException {
```

Comment:
I prefer to make classes have the following structure
```
class Class {
  // constants
  
  // fields
  
  // constructur
  
  // public methods
  
  // protected methods

  // private methods
}
```
I'd sort methods here.

It's not a requirement but just my personal taste for code =) 
So you can leave without changes if you think if it is better.

Reply:
Done in commit https://github.com/eclipse/che/pull/8805/commits/c1d9dfcbdb7f856774e1d7cf24b4069d6134bf98